src/job_processor.erl

```diff
@@ -0,0 +1,94 @@
+-module(job_processor).
+
+-define(NOTEST, true).
+-ifdef(TEST).
+    -compile(export_all).
+    -compile(nowarn_export_all).
+-endif.
+
+-type task_name()   :: binary().
+-type reply_mode()  :: sorted |  bash_script.
+
+-type task()        :: #{
+                        name := task_name(),
+                        command := binary(),
+                        requires => [task_name()]
+                    }.
+
+-export([process_tasks/2]).
+
+-export_type([reply_mode/0]).
+
+-spec process_tasks(JsonAsBinary, Mode) -> Result when
+    JsonAsBinary    :: binary(),
+    Mode            :: reply_mode(),
+    Result          :: {ok, binary()} | {error, term()}.
+
+process_tasks(JsonAsBinary, Mode) ->
+    case thoas:decode(JsonAsBinary, #{keys => to_existing_atom}) of
+        {ok, #{tasks := Tasks}} ->
+            produce_output(Mode, sort_tasks(Tasks));
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+% =========================== Private helpers ==================================
+
+-spec produce_output(Mode, SortedTasks) -> Result when
+    Mode        :: reply_mode(),
+    SortedTasks :: [task()],
+    Result      :: {ok, binary()} | {error, term()}.
+
+produce_output(sorted, SortedTasks) ->
+    {ok, thoas:encode(#{tasks => SortedTasks})};
+
+produce_output(bash_script, SortedTasks) ->
+    {ok, lists:foldl(fun(#{command := Command}, Acc) ->
+        <<Acc/binary, Command/binary, "\n">>
+    end, <<"#!/usr/bin/env bash\n">>, SortedTasks)}.
+
+-spec sort_tasks(Tasks) -> Result when
+    Tasks       :: [task()],
+    Result      :: [task()].
+
+sort_tasks(Tasks) ->
+    KeyedTasks = keyed(Tasks),
+    {Sorted, _} = lists:foldl(fun
+        (T, {Acc, Seen}) ->
+            sort(T, KeyedTasks, Acc, Seen)
+        end, {[], []}, Tasks),
+    lists:reverse(Sorted).
+
+-spec keyed(Tasks) -> Result when
+    Tasks       :: [task()],
+    Result      :: #{task_name() => task()}.
+
+keyed(Tasks) ->
+    lists:foldl(fun (#{name := Name} = Task, Acc) ->
+        maps:put(Name, Task, Acc)
+    end, #{}, Tasks).
+
+-spec sort(Task, KeyedMap, Acc, Seen) -> Result when
+    Task        :: task(),
+    KeyedMap    :: #{task_name() => task()},
+    Acc         :: [task()],
+    Seen        :: [task_name()],
+    Result      :: {task_name(), [task_name()]}.
+
+sort(#{name := Name} = Task, KeyedMap, Acc, Seen) ->
```

Comment:
also, handling cyclic dependencies is manadatory in my opinion

Reply:
@danfilip sure, let's add it:
https://github.com/spylik/cs_challenge/pull/2